pkg/ccl/backupccl/restore_test.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package backupccl
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+// A special jobs.Resumer that simulates interrupted resume by
+// aborting resume after restore descriptors are published, and then
+// resuming execution again.
+var _ jobs.Resumer = &restartAfterPublishDescriptorsResumer{}
+
+type restartAfterPublishDescriptorsResumer struct {
+	t       *testing.T
+	wrapped *restoreResumer
+}
+
+func (r *restartAfterPublishDescriptorsResumer) Resume(
+	ctx context.Context, phs interface{}, resultsCh chan<- tree.Datums,
+) error {
+	e := errors.New("bail out")
+	r.wrapped.testingKnobs.afterPublishingDescriptors = func() error {
+		return e
+	}
+	require.Equal(r.t, e, r.wrapped.Resume(ctx, phs, resultsCh))
+	r.wrapped.testingKnobs.afterPublishingDescriptors = nil
+	return r.wrapped.Resume(ctx, phs, resultsCh)
+}
+
+func (r *restartAfterPublishDescriptorsResumer) OnFailOrCancel(
+	ctx context.Context, phs interface{},
+) error {
+	return nil
+}
+
+func TestRestorePrivilegesChanged(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	dir, dirCleanupFn := testutils.TempDir(t)
+	defer dirCleanupFn()
+	params := base.TestClusterArgs{}
+	params.ServerArgs.ExternalIODir = dir
+	tc := testcluster.StartTestCluster(t, 1, params)
+	defer tc.Stopper().Stop(ctx)
+
+	tc.Server(0).JobRegistry().(*jobs.Registry).TestingResumerCreationKnobs = map[jobspb.Type]func(raw jobs.Resumer) jobs.Resumer{
```

Comment:
Nice test!

Not for this PR, but I think it would be nice to have this wrapping resumer injected in the registry during `backupRestoreTestSetupWithParams` so that we can get this goodness across all of the existing tests.

@dt also mentioned the idea of having "breakpoints" or knobs set during interesting points throughout the execution. Perhaps we could randomly select one of the breakpoints to return this error every time to get varied coverage on resuming the job at different points throughout the lifecycle of the job.

Reply:
Thanks.  I do think we should lift this case to the master.   The only thing is: i had to put it under backupccl package, and not under backupccl_test package so that I can access resumer.  The injection approach was basically copied from the import test.

Reply:
I did a refactor of the backup tests in 20.2 to move most of them from `backupccl_test` to `backupccl` (for exactly the same reason, so that I could put testing knobs on the resumer).